migration/blocktransactions/blocktransactions.go

```diff
@@ -0,0 +1,291 @@
+
+	err := deleteOldData(batch, startBlock, endBlock)
+	return batch, err
+}
+
+func debugLog(logger utils.SimpleLogger, msg string, startBlock, endBlock uint64) {
+	logger.Debugw(
+		msg,
+		"startBlock", startBlock,
+		"endBlock", endBlock,
+		"numGoroutines", runtime.NumGoroutine(),
+	)
+}
+
+func getFirstBlockInBucket[A any](items iter.Seq2[prefix.Entry[A], error]) (uint64, bool, error) {
+	for item, err := range items {
+		if err != nil {
+			return 0, false, err
+		}
+
+		var blockNumIndex db.BlockNumIndexKey
+		if err := blockNumIndex.UnmarshalBinary(item.Key[1:]); err != nil {
+			return 0, false, err
+		}
+		return blockNumIndex.Number, true, nil
+	}
+	return 0, false, nil
+}
+
+func getFirstBlock(database db.KeyValueReader) (uint64, bool, error) {
+	transactions, hasTransactions, err := getFirstBlockInBucket(
+		core.TransactionsByBlockNumberAndIndexBucket.Prefix().Scan(database),
+	)
+	if err != nil {
+		return 0, false, err
+	}
+
+	receipts, hasReceipts, err := getFirstBlockInBucket(
+		core.ReceiptsByBlockNumberAndIndexBucket.Prefix().Scan(database),
+	)
+	if err != nil {
+		return 0, false, err
+	}
+
+	if !hasReceipts && !hasTransactions {
+		return 0, false, nil
+	}
+
+	minBlock := min(transactions, receipts)
+	minBlock -= minBlock % batchSize
+	return minBlock, true, nil
+}
+
+func clearOldBuckets(database db.KeyValueStore) error {
+	err := core.TransactionsByBlockNumberAndIndexBucket.Prefix().DeletePrefix(database)
+	if err != nil {
+		return err
+	}
+
+	err = core.ReceiptsByBlockNumberAndIndexBucket.Prefix().DeletePrefix(database)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func migrateBlockRange(
+	ctx context.Context,
+	database db.KeyValueStore,
+	logger utils.SimpleLogger,
+	cleanup *cleanup,
+	firstBlock,
+	chainHeight uint64,
+) []byte {
+	counter := newCounter(logger, timeLogRate)
+
+	s := stream.New().WithMaxGoroutines(batchConcurrency)
+	defer s.Wait()
+
+	for startBlock := firstBlock; startBlock <= chainHeight; startBlock += batchSize {
+		select {
+		case <-ctx.Done():
+			return []byte{}
+		default:
+		}
+
+		s.Go(func() stream.Callback {
+			endBlock := min(startBlock+batchSize-1, chainHeight)
+			batch, err := createBatch(logger, database, startBlock, endBlock)
+
+			return func() {
+				if err != nil {
+					cleanup.AddError(err)
+					return
+				}
+
+				debugLog(logger, "writing block", startBlock, endBlock)
+				if err := batch.Write(); err != nil {
+					cleanup.AddError(err)
+					return
+				}
+
+				counter.log(batch, batchSize)
+
+				debugLog(logger, "wrote block", startBlock, endBlock)
+
+				cleanup.CloseBatch(batch)
+			}
+		})
+	}
+
+	return nil
+}
+
+func (BlockTransactionsMigration) Before([]byte) error {
+	return nil
+}
+
+func (BlockTransactionsMigration) Migrate(
+	ctx context.Context,
+	database db.KeyValueStore,
+	network *utils.Network,
+	logger utils.SimpleLogger,
+) ([]byte, error) {
+	for {
+		firstBlock, hasFirstBlock, err := getFirstBlock(database)
```

Comment:
nitpick: I think `getFirstBlock` is a bit more readable if it is called `getFirstBlockNumber`

Reply:
Updated, addressed in #3348